CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies

sdumitriu opened the pull request:
Draft PR, still left to do:

- [x] Update documentation
- [x] Update the start.sh script
- [x] Fix Users and Groups administration
- [x] Update docker and compose files
- [x] Test upgrading from an older version

Future work, out of scope, to be addressed separately:
-  Move projects out into their own repos
-  Test clustering multiple instances
-  Test upgrading from an older version
-  Figure out a way to have custom project frontend code on top of aggregated-frontend without duplicating all of the vendor scripts if possible

Starting the base CARDS:
```
./start_cards.sh
```

Starting at a different port:
```
./start_cards.sh -p 9090
```

Starting with a project:
```
./start_cards.sh -P cards4lfs
```

Starting with multiple projects:
```
./start_cards.sh -P cards4lfs,cards4proms
```

Starting with a different permission scheme:
```
./start_cards.sh --permissions trusted
```

Mapping between old runmodes and new parameters:

runMode | parameter
------------ | ---------
`oak_tar` | None, this is the default
`oak_mongo` | `--mongo`
`dev` | `--dev`
`cardiac_rehab` | `-P cards4care`
`kids` | `-P cards4kids`
`lfs` | `-P cards4lfs`
`proms` | `-P cards4proms`
`test` | `--test`
`demo` | `--demo`
`permissions_open` | `--permissions open` (or nothing, this is the default)
`permissions_trusted` | `--permissions trusted`
`permissions_ownership` | `--permissions ownership`

sdumitriu committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies
Update sling and oak, switch to feature files instead of provisioning files;
The upstream version of the feature files are committed, to be patched in next commit

sdumitriu committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies
Put back our custom settings:
- custom 503 page
- support for providing a logback.xml configuration file
- custom mongo hostname
- auto checkout and checkin
- disable self registration
- custom authentication requirements for paths
- switch default encoding to UTF8

sdumitriu committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies
Converted the provisioning files to feature files

sdumitriu committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies
Delete the new launcher data directory

sdumitriu committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies
Transformed projects into features

sdumitriu committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies
Remove old slingstart plugin

sdumitriu committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies
Updated the start.sh script to use the new java command

sdumitriu committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies
Do not include composum by default

sdumitriu committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies
- Allow backwards compatibility for the `-p` argument to specify a port
- add support for `-P` to specify a project to include
- add support for using VERSION in feature URIs and use the current version

sdumitriu committed:
[misc] Always allow unlimited entity expansion to allow installing vocabularies without errors

sdumitriu committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies
Startup script improvement: allow `--dev` as a parameter to include Composum

sdumitriu committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies
Startup script improvement: allow specifying more than one project, comma-separated, for the `-P` argument

sdumitriu committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies
Startup script improvement: allow switching to Mongo storage

sdumitriu committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies
Startup script improvement: easy permissions pick with `--permissions`

sdumitriu committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies
Minor doc improvements

sdumitriu committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies
First version of the new Dockerfile, can start an empty cards core

sdumitriu committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies
Allow specifying the mongo connection parameters using variables

sdumitriu committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies
Added missing files

sashaandjic commented:
Vocabulary search is not working.
![image](https://user-images.githubusercontent.com/23243244/137558055-797a5ef0-49cb-4f85-850b-cbb39fa1d9f2.png)
![image](https://user-images.githubusercontent.com/23243244/137558053-12e8e02d-3723-4727-8d79-6fd45a8e5557.png)

sdumitriu committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies
Bugfix: Proms patient UI failed to load

sdumitriu committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies
Bugfixes: user and group management doesn't work, vocabulary questions don't work

sdumitriu commented:
> QUESTION: no runmode for dev?

I forgot to list it, it's `--dev`.

> Vocabulary search is not working.

Fixed.

sdumitriu committed:
CARDS-1411: Non-admin users should be able to access the Proms node

sdumitriu committed:
CARDS-1346: Configure permissions to prevent patient users from accessing the computed interpretation of their answers

sdumitriu committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies
Bugfix: computed questions break form saving

sdumitriu committed:
CARDS-1346: Configure permissions to prevent patient users from accessing the computed interpretation of their answers
Use a more privileged service account to read the full questionnaire, not just the accessible questions

sdumitriu committed:
[misc] Removed leftover comment

sdumitriu committed:
CARDS-1347: Evaluate computed fields serverside
Bugfix: during startup there may be no session available, which causes NPEs and failure to properly initialize the system

sdumitriu committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies
Bugfix: wrong app name and color theme used in projects

sdumitriu committed:
[misc] Change the instance data directory to .cards-data

sashaandjic commented:
![image](https://user-images.githubusercontent.com/23243244/139550192-bc306ee0-5980-4bc0-9b17-f32bdebf2522.png)

sashaandjic commented:
Looks like filters are not working again.
![image](https://user-images.githubusercontent.com/23243244/139553481-014aec4a-b492-48c3-9571-8b0a15ad88f8.png)
![image](https://user-images.githubusercontent.com/23243244/139553557-bd79cc37-20c1-4dfa-be15-dc7643292a87.png)

sdumitriu commented:
> Looks like filters are not working again.

It seems that "decimal" is indexed like a string, not like a number, so the above range will never work since for strings "5 > 22", so it's impossible to have a value both greater than 5 and smaller than 22 when 5 is already greater than 22.

This is not a bug in 469, it's a missed usecase in 1494 (same thing happens on dev).

sashaandjic reviewed (approved):
tested

IntegralProgrammer committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies

Specify the CARDS_VERSION environment variable in the Dockerfile

IntegralProgrammer committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies

Add legacy support for the ADDITIONAL_RUN_MODES environment variable in docker_entry.sh

IntegralProgrammer committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies

Allow for the permissions mode (open, trusted, ownership) to be set through the PERMISSIONS environment variable

IntegralProgrammer committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies

Allow for the CARDS project (cards4care, cards4kids, cards4lfs, cards4proms) project to be set through the CARDS_PROJECT environment variable

IntegralProgrammer committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies

Move comment

IntegralProgrammer committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies

If the DEV environment variable is specified, make Composum available

IntegralProgrammer committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies

Allow for additional Apache Sling features to be specified through the ADDITIONAL_SLING_FEATURES environment variable

IntegralProgrammer committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies

Switch from the openjdk:8-jre-alpine Docker image which is based on Alpine Linux 3.9 (no longer supported) to the latest Alpine Linux release, 3.14

IntegralProgrammer committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies

Build all dependency JARs in the first stage of a multi-stage Docker build so that the final production Docker image may be self-contained and as slim as possible

IntegralProgrammer committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies

Only build dependency JARs in the first stage of a multi-stage Docker build if -Ddocker.buildArg.build_jars=true is specified in MAVEN_OPTS

IntegralProgrammer committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies

Only build dependency JARs in the first stage of a multi-stage Docker build if -Ddocker.buildArg.build_jars=true is specified in MAVEN_OPTS

Move the commands that are only needed for JAR building into the make_filled_m2.sh script

IntegralProgrammer committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies

Only build dependency JARs in the first stage of a multi-stage Docker build if -Ddocker.buildArg.build_jars=true is specified in MAVEN_OPTS

Have Maven automatically create the COMPLETE_SOURCE_CODE.tar archive file of the current state of the local git repository add ADD that to the first stage container of the Docker build

IntegralProgrammer committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies

Only build dependency JARs in the first stage of a multi-stage Docker build if -Ddocker.buildArg.build_jars=true is specified in MAVEN_OPTS

Use build_jars==true instead of only requiring that build_jars not be null

IntegralProgrammer committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies

For complete_git_copy.sh, if we are not inside a git repository, because we are in the Docker build container, exit with a status zero as we do not need to create the COMPLETE_SOURCE_CODE.tar archive within the build container

IntegralProgrammer committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies

Update generate_compose_yaml.py with --cards_project parameter

IntegralProgrammer committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies

Change the instance data directory to .cards-data for everything Docker-related

IntegralProgrammer committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies

Add support for an external Mongo database in docker_entry.sh

IntegralProgrammer committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies

Switch from the openjdk:8-jre-alpine Docker image which is based on Alpine Linux 3.9 (no longer supported) to the latest Alpine Linux release, 3.14

Add this change to cleanup.sh as well

IntegralProgrammer committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies

[misc] Add __pycache__ directory to the .gitignore file for the compose-cluster directory

IntegralProgrammer committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies

[misc] Clean up tabs+spaces formatting in generate_compose_yaml.py. Use two spaces for an indent

sashaandjic commented:
build is still failing
![image](https://user-images.githubusercontent.com/23243244/141533966-356cf237-b782-4cd3-9d6d-17efd2be453a.png)

sdumitriu committed:
[misc] Fixed build issue with usernames containing spaces

sashaandjic reviewed (approved):
tested

IntegralProgrammer commented:
## Testing Instructions for Docker parts for CARDS-469

- Before doing anything else, create a Docker Compose deployment of Cards4CaRe, backed by a local MongoDB server, based on the `dev` branch.
  - `docker pull mongo:4.2-bionic`
  - `docker run --rm -p 27017:27017 --name mongo -d mongo:4.2-bionic`
  - `git checkout dev`
  - `mvn clean install`
  - `cd compose-cluster`
  - `CARDS_EXT_MONGO_AUTH='' python3 generate_compose_yaml.py --external_mongo --external_mongo_address 172.99.0.1 --external_mongo_dbname sling --subnet 172.99.0.0/16`
  - Edit the generated `docker-compose.yml` file and under `services`/`cardsinitial`/`environment` add the entry `- ADDITIONAL_RUN_MODES=dev,cardiac_rehab`
  - `docker-compose build`
  - `docker-compose up -d`
  - Visit `http://localhost:8080`, login as `admin`:`admin`, create some Forms/Subjects
  - `docker-compose down`
  - `docker-compose rm`
  - `docker volume prune -f`
  - Backup this `compose-cluster` directory to somewhere outside of the _Cards_ git repository
  - `./cleanup.sh`
  - Do not stop the running Mongo DB Docker container
  - Switch to the `CARDS-469` branch (`git checkout CARDS-469`)
 
#### Can we build a self-contained _production_ Docker image that includes all required JAR dependencies?

- `MAVEN_OPTS="-Ddocker.verbose -Ddocker.buildArg.build_jars=true" mvn clean install`
  - WARNING: Running this type of build will cause all dependencies to be downloaded by Maven and therefore will take approximately 20 minutes to run and use ~800MB of bandwidth.
- Starting this image without mounting the host's `~/.m2` directory should work (`docker run --rm -p 8080:8080 -e OAK_FILESYSTEM=true -it cards/cards`)

#### Does legacy support for previous Docker Compose environments work properly?
  - Enter the backed up `compose-cluster` directory that you have created in the first step
  - Ensure that you have built the self-contained Cards Docker image from the previous step
  - Start the Docker Compose environment (`docker-compose up -d`)
  - Visit `http://localhost:8080` and login as `admin`:`admin`. You should have a Cards4CaRe instance that contains the Forms/Subjects that were previously created.
  - `docker-compose down`
  - `docker-compose rm`
  - `docker volume prune -f`
  - You may now stop the Mongo Docker container that was started in the first step (`docker stop mongo`)

#### Can we use `generate_compose_yaml.py` to provide a deployment that uses a self-provided Mongo cluster for data storage?
  - `cd compose-cluster`
  - `python3 generate_compose_yaml.py --shards 2 --replicas 3`
  - `docker-compose build`
  - `docker-compose up -d`
  - Visit `http://localhost:8080` and verify that everything works as expected
  - Verify that Cards is using this Mongo database cluster to store its data (you should see `sling` entries in the following)
  ```bash
  docker-compose exec router mongo
  sh.status()
  exit
  ```
  - `docker-compose down`
  - `docker-compose rm`
  - `docker volume prune -f`
  - `./cleanup.sh`

#### Can we use `generate_compose_yaml.py` to provide a deployment that uses an externally provided Mongo database for data storage?
  - `cd compose-cluster`
  - `docker run --rm -p 27017:27017 --name mongo -d mongo:4.2-bionic`
  - `CARDS_EXT_MONGO_AUTH='' python3 generate_compose_yaml.py --external_mongo --external_mongo_address 172.99.0.1 --external_mongo_dbname sling --subnet 172.99.0.0/16`
  - `docker-compose up -d`
  - Visit `http://localhost:8080`, login as `admin`:`admin`, and create some Forms/Subjects
  - `docker-compose down`
  - `docker-compose rm`
  - `docker volume prune -f`
  - `docker-compose up -d`
  - Visit `http://localhost:8080`, login as `admin`:`admin`, and ensure that the previously created Forms/Subjects are still there
  - `docker-compose down`
  - `docker-compose rm`
  - `docker stop mongo`
  - `docker volume prune -f`
  - `docker run --rm -p 27017:27017 --name mongo -d mongo:4.2-bionic`
  - `docker-compose up -d`
  - Visit `http://localhost:8080` and login as `admin`:`admin`. The previously created Forms/Subjects should no longer be there now that Mongo has been restarted.
  - `docker-compose down`
  - `docker-compose rm`
  - `docker stop mongo`
  - `docker volume prune -f`
  - `./cleanup.sh`

#### Can we use `generate_compose_yaml.py` to provide a deployment that uses the file system for data storage ?
  - `cd compose-cluster`
  - `python3 generate_compose_yaml.py --oak_filesystem`
  - `docker-compose up -d`
  - Visit `http://localhost:8080`. Everything should work as expected.
  - `docker-compose down`
  - `docker-compose rm`
  - `docker volume prune -f`
  - `./cleanup.sh`

#### Can permissions be set properly using environment variables in `docker-compose.yml`?
  - This is done through the `PERMISSIONS` environment variable that is passed to the `cardsinitial` Docker container. It can be set to one of; `open`, `trusted`, or `ownership`. Ensure that each one works as expected. Additionally, not setting the `PERMISSIONS` environment variable will default to the `open` permissions mode.

####  Can Composum be enabled through `docker-compose.yml`?
  - This is done through the `DEV` environment variable that is passed to the `cardsinitial` Docker container. When it is present, Composum should be available. When it is not present, and the legacy `dev` runMode is not specified in the `ADDITIONAL_RUN_MODES` environment variable, Composum should not be present.

#### Does the `--cards_project` flag in `generate_compose_yaml.py` allow for different Cards projects to be created?
  - Valid values for `--cards_project` are; `cards4care`, `cards4kids`, `cards4lfs`, `cards4proms`

#### Does the `ADDITIONAL_SLING_FEATURES` environment variable in `docker-compose.yml` work as expected?
  - By setting the `ADDITIONAL_SLING_FEATURES` environment variable for the `cardsinitial` container in `docker-compose.yml` to `mvn:io.uhndata.cards/cards-modules-demo-banner/0.9-SNAPSHOT/slingosgifeature` and starting the Docker Compose set of services, the demo banner should be present when visiting `http://localhost:8080` and logging in as `admin`.

#### Can we build a _development_ Docker Image which does _not_ include required JAR dependencies?

- `MAVEN_OPTS="-Ddocker.verbose -Ddocker.buildArg.build_jars=false" mvn clean install`
- Starting this image without mounting the host's `~/.m2` directory should fail (`docker run --rm -p 8080:8080 -e OAK_FILESYSTEM=true -it cards/cards`)
- Starting this image with the host's `~/.m2` directory mounted to the container's `/root/.m2` directory should work (`docker run --rm -p 8080:8080 -e OAK_FILESYSTEM=true -v $(realpath ~/.m2):/root/.m2:ro -it cards/cards`)

sdumitriu committed:
[misc] Fixed rebase issues

IntegralProgrammer committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies

Skip over the .cards-data directory in Utilities/Development/list_extension_points.py

IntegralProgrammer committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies

Add support for --demo and --test flags for start_cards.sh

IntegralProgrammer committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies

Add support for enabling the demo features through the DEMO environment variable and the test features through the ENABLE_TEST_FEATURES environment variable in docker_entry.sh

marta- committed:
CARDS-469: Upgrade versions of sling, jackrabbit, oak and dependencies

Updated readme with `--demo` and `--test` flags examples

sashaandjic reviewed (approved):
tested

acrowthe reviewed (approved):
Docker functionality works as expected

sashaandjic commented:
QUESTION: Is this expected?
![image](https://user-images.githubusercontent.com/23243244/144328634-0e43ec9a-ac8a-4d98-ba99-9c8802d59b91.png)
